### PR TITLE
Remove FlowProject.add_operation

### PR DESCRIPTION
**Pull request by b-butler**
## Description

<s>This PR makes `FlowProject.add_operation` a classmethod as per the development team's discussion on 2021-03-29. The method does essentially what it did before, but at the class level instead of the instance level storage. This means that using the method on an instance will not add the operation nor be capable of warning the user. A test of this method is added to `test_project.py`.</s>

I removed `FlowProject.add_operation` and added a warning in its place. This decision for direct removal comes from conversations shown in the PR and offline. We have decided to remove the registering of operation at both the instance and operation level, in favor of only registering at the class level.

In addition, I noticed some API requirements that we were not enforcing or were not enforcing early in terms of operation functions/names which I added checks for. I can add tests for these if others like the new checks.

## Motivation and Context

We decided to change `FlowProject.add_operation` in the development meeting on 2021-03-29 to address #479.

## Types of Changes

- [ ] Documentation update
- [x] Bug fix
- [ ] New feature
- [x] Breaking change<sup>1</sup>

<sup>1</sup>The change breaks (or has the potential to break) existing functionality.

## Checklist:

- [x] I am familiar with the [**Contributing Guidelines**](https://github.com/glotzerlab/signac-flow/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**Contributor Agreement**](https://github.com/glotzerlab/signac-flow/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/signac-flow/blob/master/contributors.yaml).
- [x] My code follows the [code style guideline](https://github.com/glotzerlab/signac-flow/blob/master/CONTRIBUTING.md#code-style) of this project.
- [x] The changes introduced by this pull request are covered by existing or newly introduced tests.

If necessary:
- [x] I have updated the API documentation as part of the package doc-strings.
- [ ] I have created a separate pull request to update the [framework documentation](https://docs.signac.io/) on [signac-docs](https://github.com/glotzerlab/signac-docs) and linked it here.
- [x] I have updated the [changelog](https://github.com/glotzerlab/signac-flow/blob/master/changelog.txt).

### CODE REVIEW

**Comment by b-butler** [comment]
I am actually not sure if it makes sense now for `add_operation` to be a classmethod as it essentially just does what the decorators do a bit more concisely in some cases. I would be happy to field others' thoughts on this.

**Review by bdice — CHANGES_REQUESTED** [review]
I chatted with @b-butler offline about his comment above (whether it makes sense to keep `add_operation`). The changes in this PR are currently *incompatible* with the past instance-level behavior of `add_operation`. That is, existing projects using `add_operation` would need to be updated in order to work (registering an operation with the class would require a re-instantiation of the class to get a new instance with those operations defined). The new behavior is also somewhat redundant with the class-level registrations that are performed by `@FlowProject.operation` decorator syntax.

I would like to get input from @glotzerlab/signac-committers on whether it is important to us to keep supporting instance-level operations and/or the `add_operation` style syntax in general. I do not think the benefits in functionality outweigh the costs in engineering time and future maintenance. In addition, the current implementations are a source of significant complexity in the codebase:
- #479 arose because this API was untested (a sign of low usage)
- Resolving #237, #309, and several other issues (can't locate the one I'm thinking of at the moment) would be easier if the code base didn't have to support so many methods of interaction/registration

I vote to remove this API entirely, with no deprecation period (since it is already broken).

**Comment by csadorf** [comment]
> I would like to get input from @glotzerlab/signac-committers on whether it is important to us to keep supporting instance-level operations and/or the `add_operation` style syntax in general. I do not think the benefits in functionality outweigh the costs in engineering time and future maintenance. In addition, the current implementations are a source of significant complexity in the codebase:

While unfortunate, I agree that it appears that the effort for maintaining this API seems to be too high at this point and removal is warranted. It is especially unfortunate that the removal would go via the route *untested -> broken -> removed* without any kind of controlled deprecation. I am still a bit surprised that it is difficult to maintain it as it is essentially a wrapper at this point, but I admit that I did not have a chance yet to look at the technical details of the problem. As a mitigation measure, I suggest we provide an extremely clear error message when people use this function after all with an explanation as to 1. why the function was removed, 2. why so sudden, and 3. how to migrate.

As one alternative to immediate removal: Could we provide a minimal variant of this function that is not functionally equivalent to `@operation`, but provides some basic support? That would give us a chance to properly deprecate and remove it.

**Comment by vyasr** [comment]
> I chatted with @b-butler offline about his comment above (whether it makes sense to keep `add_operation`). The changes in this PR are currently _incompatible_ with the past instance-level behavior of `add_operation`. That is, existing projects using `add_operation` would need to be updated in order to work (registering an operation with the class would require a re-instantiation of the class to get a new instance with those operations defined). The new behavior is also somewhat redundant with the class-level registrations that are performed by `@FlowProject.operation` decorator syntax.
> 
> I would like to get input from @glotzerlab/signac-committers on whether it is important to us to keep supporting instance-level operations and/or the `add_operation` style syntax in general. I do not think the benefits in functionality outweigh the costs in engineering time and future maintenance. In addition, the current implementations are a source of significant complexity in the codebase:
> 
>     * #479 arose because this API was untested (a sign of low usage)
> 
>     * Resolving #237, #309, and several other issues (can't locate the one I'm thinking of at the moment) would be easier if the code base didn't have to support so many methods of interaction/registration
> 
> 
> I vote to remove this API entirely, with no deprecation period (since it is already broken).

Clearly I didn't think through this suggestion when I made it enough to recognize that at the point `add_operation` is called the object is already instantiated and therefore has already collected the class-level operations. Sorry about making you do this work to figure that out @b-butler. I think fixing this problem is certainly possible, but quite a bit more work than it's worth and am also in favor of removing the old API.

> While unfortunate, I agree that it appears that the effort for maintaining this API seems to be too high at this point and removal is warranted. It is especially unfortunate that the removal would go via the route _untested -> broken -> removed_ without any kind of controlled deprecation. I am still a bit surprised that it is difficult to maintain it as it is essentially a wrapper at this point, but I admit that I did not have a chance yet to look at the technical details of the problem. As a mitigation measure, I suggest we provide an extremely clear error message when people use this function after all with an explanation as to 1. why the function was removed, 2. why so sudden, and 3. how to migrate.

@csadorf here's a summary of why maintaining `add_operation` is hard now. With the introduction of the decorator syntax back in 0.6 or so, we created an independent code path because some operations were registered to `FlowProject` subclasses (via the decorator) while other operations were added upon instance creation (using `add_operation`). That was largely handled using some registration logic that, upon instantiation of a `FlowProject` or a subclass, would 1) walk the MRO and pull all operations associated with all parent classes, then 2) add the operations created via `add_operation` (which in theory could happen any time after instantiation, but in practice was done in `__init__` because everyone copied that pattern from early examples of the project template etc). That was largely manageable until the introduction of groups, but now between groups and aggregation there are a lot of different entities that have to be managed independently along these code paths then merged on project instantiation, and that's what makes it nontrivial to maintain. There's much longer discussion of this on #382 if you'd like to look through that and see if it sparks any ideas on a better solution.

At this stage @b-butler and myself are probably the only ones currently involved with the project who are intimately familiar with the code paths taken using the old API and how they interact with the decorator API. You and @b-butler worked very closely together on groups, so you made sure that that groups would support the old API, but the lack of examples or test using the old API allowed @kidrahahjo (who's done a fantastic job of it) to implement all of aggregation essentially without considering the implications for the old API. I completely dropped the ball on considering this when reviewing aggregation work, but also considering how long aggregation has taken to get finished (and considering that groups took even longer!) I'm not sure that this is a bad thing in hindsight. The upshot is that it probably would have required investing significantly more dev effort for minimal payoff. So while I agree that the current situation is unfortunate and would have preferred a proper deprecation cycle, I think a loud error message is probably the best option at this stage.

**Comment by bdice** [comment]
@vyasr I was hunting all over for #382, that's the issue I couldn't find! 😄

As for strategies to move forward, it sounds like "loud error message" is what we're agreeing on for now. I think I would implement this as:

```python
def add_operation(self, *args, **kwargs):
    raise AttributeError(
		"The add_operation method has been removed without deprecation. This "
		"message will be removed in version 0.16."
	)
```

Then we can start to clean up the messy registration logic.

**Comment by vyasr** [comment]
Probably because you were looking in issues, but #382 was a PR 😄 (which you opened while you were trying to make what you thought was a minor fix 😂 if only). @csadorf in particular, what you would probably want to read are [@bdice's summary of what he found and his questions about the internals of operation/groups](https://github.com/glotzerlab/signac-flow/pull/382#issuecomment-739456466), and [my subsequent comment on that thread](https://github.com/glotzerlab/signac-flow/pull/382#issuecomment-739679375) explaining why we have these two paths. Note that the discussion on #382 is only part of the equation because it doesn't account for the ways in which aggregations are mapped to groups (which also would need to support both the old and new APIs.

**Comment by csadorf** [comment]
Yes, we should completely remove any duplication of internal logic which also includes any registration of operations after instantiation, this is something we should simply no longer support. What I'm still wondering is whether we could at least provide a very simple wrapper classmethod as a mitigation measure.

I just implemented the following snippet:
```python
class Project(FlowProject):

    @classmethod
    def add_operation(cls, name, operation, directives=None, pre=None, post=None):

        if isinstance(operation, str):

            @flow.cmd
            def op_(job):
                return f"{operation} {job}"

            op_ = cls.operation(op_, name=name)

        else:
            op_ = cls.operation(operation, name=name)

        if directives is not None:
            op_ = flow.directives(**directives)(op_)

        if pre:
            for cond in pre:
                op_ = cls.pre(cond)(op_)
        if post:
            for cond in post:
                op_ = cls.post(cond)(op_)

        return op_
```

This actually works so I am bit confused as to what the issue is.

**Comment by bdice** [comment]
@csadorf I think the `add_operation` API is currently usable like this:

```python
project = Project()
project.add_operation(...)
project.main()
```
In my understanding, the snippet you're proposing is similar to what's in this PR (unless I'm not seeing some crucial difference) -- but this won't work on instances in the same way that the old instance method could.

**Comment by csadorf** [comment]
> @csadorf I think the `add_operation` instance method API is meant to be usable like this:
> 
> ```python
> project = Project()
> project.add_operation(...)
> project.main()
> ```
> 
> In my understanding, the snippet you're proposing is similar to what's in this PR (unless I'm not seeing some crucial difference) -- but this won't work on instances in the same way that the old instance method could. The operation will be registered with the class `Project` but the instance `project` already exists and won't find that operation unless re-instantiated. I'm not 100% sure if that is correct but that's my understanding of the current implementation.

You're right, I should have read the patch more diligently. I don't mean to imply that this would work without any changes, on the user side, but it might make the migration a bit easier. But either way, I am also ok with just removing it.

Maybe we can make the error message a bit nicer? How about this:

```python
def add_operation(self, *args, **kwargs):
    raise AttributeError(
        "The add_operation() method was removed in version 0.14. "
        "Please see https://docs.signac.io/en/latest/flow-project.html#defining-a-workflow "
        "for instructions on how to define operations using the current API. This message "
        "will be removed in version 0.16."
    )
```

I don't think we actually need to mention that it was removed without deprecation; it just adds insult to injury...

**Comment by bdice** [comment]
@csadorf I like your message! Thanks for the suggestion. 😊

**Comment by b-butler** [comment]
@bdice and @csadorf, I have removed the method and left the suggested error message.

File: flow/project.py
Comment (bdice) [inline flow/project.py:4023]: The structure of `x` and meaning of `x[0]` are not clear from this pattern. We can avoid the map/lambda and use readable names. Here are two suggestions, pick whichever you prefer.
```suggestion
        if any(name == op_name for op_name, op_func in cls._OPERATION_FUNCTIONS):
            raise ValueError(f"An operation with name '{name}' is already registered.")
        if any(func == op_func for op_name, op_func in cls._OPERATION_FUNCTIONS):
            raise ValueError("An operation with this function is already registered.")
```
Or iterate just one time:
```suggestion
        for op_name, op_func in cls._OPERATION_FUNCTIONS:
            if name == op_name:
                raise ValueError(f"An operation with name '{name}' is already registered.")
            elif func == op_func:
                raise ValueError("An operation with this function is already registered.")
```

**Review by bdice — APPROVED** [review]
@b-butler I pushed a couple changes for cleanup and added one suggestion for improvement. I approve this after that suggestion is resolved.

**Comment by bdice** [comment]
@b-butler Can you update this PR description to match its current implementation, for future reference? It can be brief.

File: flow/project.py
Comment (csadorf) [inline flow/project.py:3937]: Does it make sense to use the `@deprecated` decorator here since the function was not deprecated, but just removed?

File: flow/project.py
Comment (csadorf) [inline flow/project.py:4049]:
```suggestion
            if func is registered_func:
```
This is an identity check, right?

File: flow/project.py
Comment (csadorf) [inline flow/project.py:4040]: These changes seem somewhat unrelated and are not covered by the changelog entry. I assume this is in here due to historical reasons? I might suggest to split this off cleanly.

File: flow/project.py
Comment (b-butler) [reply flow/project.py:4049]: For function objects, I believe so.

File: flow/project.py
Comment (b-butler) [reply flow/project.py:4040]: We can do that, I didn't mention this in the changelog because we already make these API promises, and there is not an issue relating to this.

File: flow/project.py
Comment (bdice) [reply flow/project.py:3937]: @csadorf The function hasn't been removed -- its _implementation_ has been removed. This approach allows us to raise an `AttributeError` with a helpful message, until we remove the function entirely. The deprecation warning is possibly redundant because that user message contains the same info, but I thought it was better to include because developers might search the codebase for `deprecated` when finding code to remove around release time.

File: flow/project.py
Comment (bdice) [reply flow/project.py:4040]: Sure, let's just split these changes into a separate PR for cleanliness.

**Comment by csadorf** [comment]
@b-butler Please, re-request the review from me when this is ready. Thank you.

**Review by csadorf — APPROVED** [review]
Great!

**Pull request merged**
